### ISSUE TITLE
Help with authentication exception

### ISSUE THREAD

**Issue by judahnator** [googleads/google-ads-php#264]
I am working to get a product integrated with the Google Ads [Conversion Actions](https://developers.google.com/google-ads/api/docs/conversions). We are attempting to add new conversion actions via the Google Ads API (v2).

We have followed the instructions for such an action, as [documented here](https://developers.google.com/google-ads/api/docs/samples/add-conversion-action#php).
Of course our code looks a little different, so here is an excerpt of it:

```php
// An oAuth token previously obtained. We do have a refresh token.
// This token has been granted the following scope: 
// https://www.googleapis.com/auth/adwords
// as documented at: 
// https://developers.google.com/adwords/api/docs/guides/authentication
$oAuth2Credential = (new OAuth2TokenBuilder())
    ->withClientId('REDACTED')
    ->withClientSecret('REDACTED')
    ->withRefreshToken('REDACTED')
    ->build();

// The ID of the client website
// The user granting us their oAuth token has full access to this property 
$customerId = 'xxx-xxx-xxxx';

$googleAdsClient = (new GoogleAdsClientBuilder())

     // This developer token was obtained from our MCC account
     // retrieved from our account, as documented here:
     // https://developers.google.com/google-ads/api/docs/first-call/dev-token
    ->withDeveloperToken('REDACTED')
    ->withLoginCustomerId($customerId)

    ->withOAuth2Credential($oAuth2Credential)
    ->build();

$conversionAction = (new ConversionAction())
    ->setName((new StringValue())->setValue('TEST LABEL'))
    ->setCategory(ConversionActionCategory::PBDEFAULT)
    ->setType(ConversionActionType::WEBPAGE)
    ->setStatus(ConversionActionStatus::ENABLED)
    ->setValueSettings(
        (new ValueSettings())
            ->setAlwaysUseDefaultValue((new BoolValue())->setValue(true))
            ->setDefaultValue((new DoubleValue())->setValue(0.0))
    );

// Issues a mutate request to add the conversion action.
$conversionActionServiceClient = $googleAdsClient->getConversionActionServiceClient();
$response = $conversionActionServiceClient->mutateConversionActions(
    $customerId,
    [(new ConversionActionOperation())->setCreate($conversionAction)]
);
// That causes the exception we are seeing
```

At that last function call, we get an authentication exception that looks like this:
```
Google\ApiCore\ApiException  : {
    "message": "Request had insufficient authentication scopes.",
    "code": 7,
    "status": "PERMISSION_DENIED",
    "details": []
}
```

Now as far as we can tell, all the pieces are properly authorized. Any pointers? Anything obvious we are doing wrong?

**Comment by chiragvels** [comment]
Hi,

Upvote for this.

I was also facing the same issue, while fetching the accounts.

```
try {
                        $googleAdsServiceClient = $googleAdsClient->getGoogleAdsServiceClient();
                        $query = "SELECT customer.time_zone, "
                                . "customer.currency_code, "
                                . "customer.descriptive_name, "
                                . "customer.manager, "
                                . "customer.id, "
                                . "customer.test_account, "
                                . "customer.resource_name "
                                . "FROM customer ";
                        $response = $googleAdsServiceClient->search($customerId['customer'], $query, ['pageSize' => self::PAGE_SIZE]);

                        foreach ($response->iterateAllElements() as $googleAdsRow) {
                            $customer = $googleAdsRow->getCustomer();

                            $data[] = array('timeZone' => $customer->getTimeZoneUnwrapped(),
                                'currency_code' => $customer->getCurrencyCodeUnwrapped(),
                                'isManager' => ($customer->getManagerUnwrapped() !== NULL) ? $customer->getManagerUnwrapped() : 0,
                                'descriptive_name' => ($customer->getDescriptiveNameUnwrapped() !== NULL) ? $customer->getDescriptiveNameUnwrapped() : '',
                                'customerId' => $customer->getIdUnwrapped(),
                                'test_account' => ($customer->getTestAccountUnwrapped() !== NULL) ? $customer->getTestAccountUnwrapped() : 0,
                                'refresh_token' => $refreshToken,
                                'client_email' => $token_email
                              );
                        }
                    } catch (ApiException $e) {
                        throw $e;
                    }
```

**Comment by fiboknacky** [comment]
@chiragvels @judahnator   Is this happening recently?

**Comment by chiragvels** [comment]
Yes, It is happened recently.
Atleast I have faced not before,
My case is client had given access which can be seen under my MCC account email, but it is not part of MCC account.

**Comment by fiboknacky** [comment]
The issue looks more related to the Google Ads API server, not the client library.
Could you guys start a thread in the [Google Ads API forum](https://groups.google.com/g/adwords-api)?

**Comment by judahnator** [comment]
> Could you guys start a thread in the Google Ads API forum?

Sure. The link provided sends me to a 404 page though.

**Comment by fiboknacky** [comment]
Sorry, I updated the link.

**Comment by judahnator** [comment]
The folks over on that forum [are asking](https://groups.google.com/d/msg/adwords-api/hjGqvgjxir0/u4gBxpsUBAAJ) for me to provide the API call I am making. Is there a "debug mode" in this library to help troubleshoot?

**Comment by fiboknacky** [comment]
Please turn on [logging](https://developers.google.com/google-ads/api/docs/client-libs/php/logging) and provide the information that you've been asked.

**Comment by judahnator** [comment]
I added the following:

```php
$googleAdsClient = (new \Google\Ads\GoogleAds\Lib\V2\GoogleAdsClientBuilder())
  ->withLogger(app(\Psr\Log\LoggerInterface::class))
  ->withLogLevel(\Psr\Log\LogLevel::DEBUG)
  // ... snip
```

This is what I got in my log:

```
[2020-02-12 15:39:45] local.ERROR: {
    "message": "Request had insufficient authentication scopes.",
    "code": 7,
    "status": "PERMISSION_DENIED",
    "details": []
} {"exception":"[object] (Google\\ApiCore\\ApiException(code: 7): {
    \"message\": \"Request had insufficient authentication scopes.\",
    \"code\": 7,
    \"status\": \"PERMISSION_DENIED\",
    \"details\": []
} at /redacted/vendor/google/gax/src/ApiException.php:139)
```
This is closer, but I seem to be missing the API calls being made. Is there any way to get the API calls and response?
I know in the other [google/apiclient](https://github.com/googleapis/google-api-php-client) library its possible to log all of the "service calls."

**Comment by fiboknacky** [comment]
Thanks for trying. Could you please continue communicating with the agent on the forum? Thanks.

**Comment by fiboknacky** [comment]
@judahnator I want to follow up if you are able to get help from the forum yet?

**Comment by judahnator** [comment]
Hello @fiboknacky! Yes and no.

My company is in contact with a Google employee in private while we look deeper into this issue. We have not yet been able to pin down exactly what the problem is. Lets keep this issue open for a bit just in case the problem is with the library.

One question that might be relevant, looking at the [OAuth2TokenBuilder code here](https://github.com/googleads/google-ads-php/blob/v3.0.0/src/Google/Ads/GoogleAds/Lib/OAuth2TokenBuilder.php#L179-L183) (and the validation [here](https://github.com/googleads/google-ads-php/blob/v3.0.0/src/Google/Ads/GoogleAds/Lib/OAuth2TokenBuilder.php#L208-L222)), it looks like it does not allow you to pass in any scopes (defined as the first parameter of `UserRefreshCredentials`):
```
return new UserRefreshCredentials(null, [
    'client_id' => $this->clientId,
    'client_secret' => $this->clientSecret,
    'refresh_token' => $this->refreshToken
]);
```

Since we really have nothing to lose, and since the problem seems to be scope-related, we are trying to create a new `UserRefreshCredentials` manually passing in the scopes we need.
```
$oAuth2Credential = new \Google\Auth\Credentials\UserRefreshCredentials(
    'https://www.googleapis.com/auth/adsense',
    [
        'client_id' => config('services.google.client_id'),
        'client_secret' => config('services.google.client_secret'),
        'refresh_token' => 'REDACTED',
    ],
);
```

Of course it still isn't working for us, but I was wondering _is the omission of oAuth scopes in the client builder deliberate?_

**Comment by fiboknacky** [comment]
Hello,

The `scopes` property you see in that class is intended to be used with [service accounts](https://developers.google.com/identity/protocols/OAuth2ServiceAccount), not with a installed/web application flow.
That's why it's not passed with other parameters when creating `UserRefreshCredentials`.

If you wish to add more scopes when creating a new refresh token, could you try running [this example](https://github.com/googleads/google-ads-php/blob/v3.0.0/examples/Authentication/AuthenticateInStandaloneApplication.php#L66) instead?

Knack

**Comment by judahnator** [comment]
Gotcha. I appreciate the feedback. Ill keep you in the loop for whatever solution we can find!

**Comment by judahnator** [comment]
So after all the fuss, it turned out the problem was a typo in my script.

We are using the Google Services library, which packs a bunch of constants with it for the various authentication scopes. When setting the scopes, I saw that there was only one scope constant related to Ads and I picked that without thinking. After all, if there is only one option then it must be the correct one, right?

Well it turned out that the scope they defined was for AdSense, and that there was no pre-defined scope for AdWords. So the whole time it was actually a scope issue, and my script was doing exactly what I told it to and not what I wanted it to.

My apologies for taking up your guys time!
